events/events.go

```diff
@@ -0,0 +1,197 @@
+package events
+
+import (
+	"errors"
+	"sync"
+)
+
+var errClosed = errors.New("publisher is closed")
+
+// GetFn returns the event
+type GetFn func() interface{}
+
+// Event is an event
+type Event interface {
+	// OnGet sets the function to be executed when Get() is called
+	OnGet(GetFn)
+	// Get the latest Event
+	Get() interface{}
+}
+
+// Publisher is a publisher for an event
+type Publisher interface {
+	Event
+
+	// GetAndSubscribe returns the latest Event and a subscriber
+	GetAndSubscribe() (interface{}, Subscriber, error)
+	// NotifyAll notifies all subscribers
+	NotifyAll() error
+	// SubscriberLen returns the number of subscribers
+	SubscriberLen() int
+	// Close shuts down the publisher
+	Close()
+}
+
+// NewPublisher returns a new event publisher
+func NewPublisher() Publisher {
+	return &publisher{}
+}
+
+// Subscriber is a subscriber to an event
+type Subscriber interface {
+	Event
+
+	// Notifications returns the channel of notifications
+	Notifications() <-chan struct{}
+	// OnClose registers a function to be called on Close()
+	OnClose(func())
+	// Close closes the subscriber
+	Close()
```

Comment:
Could do however it will leak if you have some long running process that wakes up, consumes and then doesn't call `Close()` at the end up receiving updates it cares about.